packages/gatsby/src/utils/webpack.config.js

```diff
@@ -423,6 +414,28 @@ module.exports = async (
       ],
     }
 
+    // if react-hot-reloading
+    if (
+      stage === `develop` &&
+      process.env.GATSBY_HOT_LOADER !== `fast-refresh`
+    ) {
+      resolve.alias[`react-hot-loader`] = path.dirname(
+        require.resolve(`react-hot-loader/package.json`)
+      )
+
+      // We need to add @hot-loader/react-dom to support hot-loader work with hooks
+      try {
+        resolve.alias[`react-dom/server$`] = require.resolve(`react-dom/server`)
+        resolve.alias[`react-dom`] = require.resolve(`@hot-loader/react-dom`)
+      } catch (err) {
+        const { version: reactDomVersion } = require(`react-dom/package.json`)
+        const { major, minor } = semver.parse(reactDomVersion)
+        console.warn(
+          `React-Hot-Loader: please install "@hot-loader/react-dom@${major}.${minor}" to make sure all features of React are working.`
```

Comment:
`patch` part doesn't need to match?

Reply:
it's not available for every patch version of react so that's the reason why I didn't do it.

Reply:
What will happen when there is version mismatch between `react` and `@hot-loader/react-dom` (given that patch is not available for everything, I guess we could limit question to different minor versions)?

I guess the behaviour itself is undefined - but would users see any warnings about this (that's going from the messaging here that major and minor should be the same for those 2 packages)?

Reply:
I've added an extra warning to the code, in my tests if your version is higher I haven't found any immediate issue but of course, there are probably some subtle changes.